x-pack/plugins/beats/server/lib/domains/tokens.ts

```diff
@@ -26,32 +27,90 @@ export class CMTokensDomain {
   }
 
   public async getEnrollmentToken(enrollmentToken: string) {
-    return await this.adapter.getEnrollmentToken(enrollmentToken);
+    const fullToken = await this.adapter.getEnrollmentToken(enrollmentToken);
+
+    if (!fullToken) {
+      return {
+        token: null,
+        expired: true,
+      };
+    }
+
+    const { verified, expired } = this.verifyToken(
+      enrollmentToken,
+      fullToken.token || ''
+    );
+
+    if (!verified) {
+      return {
+        expired,
+        token: null,
+      };
+    }
+
+    return { ...fullToken, expired };
   }
 
   public async deleteEnrollmentToken(enrollmentToken: string) {
     return await this.adapter.deleteEnrollmentToken(enrollmentToken);
   }
 
-  public areTokensEqual(token1: string, token2: string) {
+  public verifyToken(recivedToken: string, token2: string) {
+    let tokenDecoded = false;
+    let expired = false;
+    const enrollmentTokenSecret = this.framework.getSetting(
+      'xpack.beats.encryptionKey'
+    );
+
+    try {
+      verifyToken(recivedToken, enrollmentTokenSecret);
+      tokenDecoded = true;
+    } catch (err) {
+      if (err.name === 'TokenExpiredError') {
+        expired = true;
+      }
+      tokenDecoded = false;
+    }
+
     if (
-      typeof token1 !== 'string' ||
+      typeof recivedToken !== 'string' ||
       typeof token2 !== 'string' ||
-      token1.length !== token2.length
+      recivedToken.length !== token2.length
     ) {
       // This prevents a more subtle timing attack where we know already the tokens aren't going to
       // match but still we don't return fast. Instead we compare two pre-generated random tokens using
       // the same comparison algorithm that we would use to compare two equal-length tokens.
-      return timingSafeEqual(
-        Buffer.from(RANDOM_TOKEN_1, 'utf8'),
-        Buffer.from(RANDOM_TOKEN_2, 'utf8')
-      );
+      return {
+        expired,
+        verified:
+          timingSafeEqual(
+            Buffer.from(RANDOM_TOKEN_1, 'utf8'),
+            Buffer.from(RANDOM_TOKEN_2, 'utf8')
+          ) && tokenDecoded,
+      };
     }
 
-    return timingSafeEqual(
-      Buffer.from(token1, 'utf8'),
-      Buffer.from(token2, 'utf8')
+    return {
+      expired,
+      verified:
+        timingSafeEqual(
+          Buffer.from(recivedToken, 'utf8'),
+          Buffer.from(token2, 'utf8')
+        ) && tokenDecoded,
+    };
+  }
+
+  public generateAccessToken() {
+    const enrollmentTokenSecret = this.framework.getSetting(
+      'xpack.beats.encryptionKey'
     );
+
+    const tokenData = {
+      created: moment().toJSON(),
+      randomHash: uuid.v4().replace(/-/g, ''),
```

Comment:
Could we encapsulate the code that creates this hash in a single place?

Reply:
👍